### PR TITLE
Add new simple array map that reads mapping from file

### PR DESCRIPTION
**Pull request by stwhite91**

### CODE REVIEW

**Comment by evan-charmworks** [comment]
Could you add a test case?

**Review by nitbhat — DISMISSED** [review]
Looks good

**Comment by evan-charmworks** [comment]
I was thinking the test case should check whether AMPI_MY_WTH is the expected value when using the mapfile.

File: src/ck-core/cklocation.ci [hunk 1]
```diff
@@ -49,6 +49,10 @@ module CkLocation {
     entry ReadFileMap(void);
   };
 
+  group [migratable] Simple1DReadFileMap : DefaultArrayMap {
```
[the remark this answers is not in the document]
Comment (stwhite91) [reply src/ck-core/cklocation.ci]: Done

**Comment by stwhite91** [comment]
Ready for review

File: tests/ampi/migration/test.C [hunk 2]
```diff
@@ -1,63 +1,92 @@
...
-	printf("Leaving TCHARM_Migrate_to, "
-               "FEM_My_partition is %d, "
-	       "CkMyPe() is %d, migrate_test is %d\n",
-	       rank, CkMyPe(), migrate_test);
-	//fflush(stdout);
+        int dest_pe = (my_pe + 1) % num_pes;
+	printf("Trying to migrate rank %d from PE %d to %d\n", rank, my_pe, dest_pe);
+	AMPI_Migrate_to_pe(dest_pe);
+        my_pe = get_my_pe();
+	printf("After migration of rank %d to PE %d\n", rank, my_pe);
```
Comment (evan-charmworks) [inline tests/ampi/migration/test.C:74]: whitespace issue here

File: tests/ampi/migration/test.C [hunk 3]
```diff
@@ -1,63 +1,92 @@
...
+  MPI_Init(&argc, &argv);
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+  MPI_Comm_size(MPI_COMM_WORLD, &p);
+  int my_pe = get_my_pe();
 
-  MPI_Init( &argc, &argv );
-  MPI_Comm_rank( MPI_COMM_WORLD, &rank);
-  MPI_Comm_size( MPI_COMM_WORLD, &p );
+  // Test that mapfile is being used correctly:
+  if (argc == 2) {
```
Comment (evan-charmworks) [inline tests/ampi/migration/test.C:39]: Shouldn't this be `(argc >= 2 && !strcmp(argv[1], "--test_mapfile"))`?

File: tests/ampi/migration/test.C [hunk 4]
```diff
@@ -1,63 +1,92 @@
...
+#include <vector>
+#include <cassert>
 #include "mpi.h"
-#include "charm.h" /* For CkAbort */
 
+int get_my_pe() {
+  int flag;
+  int *my_wth;
+  MPI_Comm_get_attr(MPI_COMM_WORLD, AMPI_MY_WTH, &my_wth, &flag);
+  assert(flag == 1);
```
Comment (evan-charmworks) [inline tests/ampi/migration/test.C:15]: Is it okay to call assert from MPI programs, or is it more correct to call MPI_Abort if the test fails?

Comment (stwhite91) [reply tests/ampi/migration/test.C]: MPI_Abort is better / more graceful. Changed.

File: tests/ampi/migration/test.C (hunk 3 above)
Comment (stwhite91) [reply tests/ampi/migration/test.C]: Fixed.

File: tests/ampi/migration/test.C (hunk 2 above)
Comment (stwhite91) [reply tests/ampi/migration/test.C]: Fixed.

File: tests/ampi/migration/test.C [hunk 5]
```diff
@@ -1,63 +1,106 @@
...
+  MPI_Comm_size(MPI_COMM_WORLD, &p);
+  int my_pe = get_my_pe();
+
+  // Test that mapfile is being used correctly:
+  if (argc >= 2 && !strcmp(argv[1], "--test_mapfile")) {
+    if (rank == 0) {
+      printf("Testing mapfile correctness\n");
+      FILE *mapf = fopen("mapfile", "r");
+      if (mapf == NULL) {
+	printf("Missing file named 'mapfile'!\n");
```
Comment (evan-charmworks) [inline tests/ampi/migration/test.C:49]: whitespace

File: tests/ampi/migration/test.C [hunk 6]
```diff
@@ -1,63 +1,106 @@
...
+        MPI_Abort(MPI_COMM_WORLD, 1);
+      }
 
-  int rank;            /* process id */
-  int p;                /* number of processes */
+      std::vector<int> init_pes(p);
+      for (int i=0; i<p; i++) {
+        if (fscanf(mapf, "%d\n", &init_pes[i]) != 1) {
+	  printf("Unrecongized mapfile formatting!\n");
```
Comment (evan-charmworks) [inline tests/ampi/migration/test.C:56]: whitespace

File: tests/ampi/migration/test.C [hunk 7]
```diff
@@ -1,63 +1,106 @@
...
-	//fflush(stdout);
+        int dest_pe = (my_pe + 1) % num_pes;
+	printf("Trying to migrate rank %d from PE %d to %d\n", rank, my_pe, dest_pe);
+	AMPI_Migrate_to_pe(dest_pe);
+	my_pe = get_my_pe();
+	if (my_pe != dest_pe) {
+	  printf("Rank %d is on PE %d but should have migrated to PE %d!\n", rank, my_pe, dest_pe);
+          MPI_Abort(MPI_COMM_WORLD, 1);
+        }
+	printf("After migration of rank %d to PE %d\n", rank, my_pe);
```
Comment (evan-charmworks) [inline tests/ampi/migration/test.C:88]: whitespace

**Comment by evan-charmworks** [comment]
There are still tabs on lines 50, 57, 82, 83, 84, 85, 86, and 89 of test.C.

File: tests/ampi/migration/test.C [hunk 8]
```diff
@@ -1,63 +1,107 @@
...
+        printf("Missing file named 'mapfile'!\n");
+        MPI_Abort(MPI_COMM_WORLD, 1);
+      }
 
-  int rank;            /* process id */
-  int p;                /* number of processes */
+      std::vector<int> init_pes(p);
+      for (int i=0; i<p; i++) {
+        if (fscanf(mapf, "%d\n", &init_pes[i]) != 1) {
+          printf("Unrecongized mapfile formatting!\n");
```
Comment (evan-charmworks) [inline tests/ampi/migration/test.C:57]: Typo: Unrecognized

**Review by evan-charmworks — APPROVED** [review]

**Review by nitbhat — APPROVED** [review]

**Pull request merged**